plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataHandler.php

```diff
@@ -195,6 +197,137 @@ private function is_order_newer_than_post( $order ): bool {
 		return $order_modified_gmt >= $post_modified_gmt;
 	}
 
+	/**
+	 * Builds an array with properties and metadata for which HPOS and post record have different values.
+	 * Given it's mostly informative nature, it doesn't perform any deep or recursive searches and operates only on top-level properties/metadata.
+	 *
+	 * @since 8.6.0
+	 *
+	 * @param int $order_id Order ID.
+	 * @return array Array of [HPOS value, post value] keyed by property, for all properties where HPOS and post value differ.
+	 */
+	public function get_diff_for_order( int $order_id ): array {
+		$diff = array();
+
+		$hpos_order = $this->get_order_from_datastore( $order_id, 'hpos' );
+		$cpt_order  = $this->get_order_from_datastore( $order_id, 'cpt' );
+
+		if ( $hpos_order->get_type() !== $cpt_order->get_type() ) {
+			$diff['type'] = array( $hpos_order->get_type(), $cpt_order->get_type() );
+		}
+
+		$hpos_meta = $this->order_meta_to_array( $hpos_order );
+		$cpt_meta  = $this->order_meta_to_array( $cpt_order );
+
+		// Consider only keys for which we actually have a corresponding HPOS column or are meta.
+		$all_keys = array_unique(
+			array_diff(
+				array_merge(
+					$this->get_order_base_props(),
+					array_keys( $hpos_meta ),
+					array_keys( $cpt_meta )
+				),
+				$this->data_synchronizer->get_ignored_order_props()
+			)
+		);
+
+		foreach ( $all_keys as $key ) {
+			$val1 = in_array( $key, $this->get_order_base_props(), true ) ? $hpos_order->{"get_$key"}() : ( $hpos_meta[ $key ] ?? null );
+			$val2 = in_array( $key, $this->get_order_base_props(), true ) ? $cpt_order->{"get_$key"}() : ( $cpt_meta[ $key ] ?? null );
+
+			// Workaround for https://github.com/woocommerce/woocommerce/issues/43126.
+			if ( ! $val2 && in_array( $key, array( '_billing_address_index', '_shipping_address_index' ), true ) ) {
+				$val2 = get_post_meta( $order_id, $key, true );
+			}
+
+			if ( $val1 != $val2 ) { // phpcs:ignore WordPress.PHP.StrictComparisons.LooseComparison
+				$diff[ $key ] = array( $val1, $val2 );
+			}
+		}
+
+		return $diff;
+	}
+
+	/**
+	 * Returns an order object as seen by either the HPOS or CPT datastores.
+	 *
+	 * @since 8.6.0
+	 *
+	 * @param int    $order_id      Order ID.
+	 * @param string $data_store_id Datastore to use. Should be either 'hpos' or 'cpt'. Defaults to 'hpos'.
+	 * @return \WC_Order Order instance.
+	 */
+	public function get_order_from_datastore( int $order_id, string $data_store_id = 'hpos' ) {
+		$data_store = ( 'hpos' === $data_store_id ) ? $this->data_store : $this->data_store->get_cpt_data_store_instance();
```

Comment:
Probably not for this PR, but maybe worth jotting down as a TODO somewhere or to open an issue for: 

Thinking about the case that we could add another data store in the future, this looks like something that should be handled centrally somewhere, say, as a static method on a parent class of the data store classes. Something that doesn't require us to know all possible data stores right here and now. 

Then again the premise of an additional data store might be 100% not a possibility, so maybe this would be an over-optimization.

Reply:
This is a good idea indeed. We did have a task for something along these lines when we started HPOS but I can't find it now so it was probably closed.
For this particular tool I think we're good because everything is literally under the `hpos` namespace, but I agree it's worth taking a look at how to handle this in an agnostic way. I'll add this to my HPOS todo list and add an issue (again).